Introduce base::Castanets::IsEnabled().

venumusham opened the pull request:
Now, we can enable "castanets" dynamically
via command-line flag at runtime.

[Normal mode]
- Disable castanets feature
- Browser : ./out/Default/chrome

[Castanets mode]
- Enable "distributed chromium" feature
- Browser : ./out/Default/chrome --enable-castanets
- Renderer : ./out/Default/chrome --type=renderer --enable-castanets=<server_IP_adress>

Signed-off-by: venu.musham <venu.musham@samsung.com>

younghajung commented:
@venumusham, please resolve conflicts to apply the patch.

Sunwoo-Nam reviewed (commented):
The later this pr is applied, the more work to be done.
Could you resolve conflicts as soon as possible?

venumusham committed:
Introduce base::Castanets::IsEnabled().

Now, we can enable "castanets" dynamically
via command-line flag at runtime.

[Normal mode]
- Disable castanets feature
- Browser : ./out/Default/chrome

[Castanets mode]
- Enable "distributed chromium" feature
- Browser : ./out/Default/chrome --enable-castanets
- Renderer : ./out/Default/chrome --type=renderer --enable-castanets=<server_IP_adress>

Signed-off-by: venu.musham <venu.musham@samsung.com>

venumusham commented:
Rebase complete, @reviewers pls review

suyambulingamrm reviewed (commented):
These commits modifies some source files..
https://github.com/Samsung/Castanets/commit/39134d45051118ff129527fbf27eeb4bf427c0ad
https://github.com/Samsung/Castanets/commit/1f536b3ee6b986225886c2e3e933a05e0871e731

Can you review if runtime flag is necessary here ?

venumusham commented:
> These commits modifies some source files..
> [39134d4](https://github.com/Samsung/Castanets/commit/39134d45051118ff129527fbf27eeb4bf427c0ad)
> [1f536b3](https://github.com/Samsung/Castanets/commit/1f536b3ee6b986225886c2e3e933a05e0871e731)
> 
> Can you review if runtime flag is necessary here ?

added runtime flag where its needed.

younghajung commented on chrome/app/chrome_main_delegate.cc:
```diff
@@ -851,11 +855,13 @@ void ChromeMainDelegate::PreSandboxStartup() {
-    ui::ResourceBundle::GetSharedInstance().AddDataPackFromAsset("assets/resources.pak");
+    if (base::Castanets::IsEnabled()) {
+      ui::MaterialDesignController::Initialize();
+      const std::string loaded_locale =
```
`loaded_locale' should be declared out of  `if (base::Castanets::IsEnabled())`.
Please see line 911.

venumusham replied:
ok, will update

younghajung commented on chrome/app/chrome_main_delegate.cc:
```diff
@@ -851,11 +855,13 @@ void ChromeMainDelegate::PreSandboxStartup() {
+          ui::ResourceBundle::InitSharedInstanceWithLocale(
+              locale, NULL, ui::ResourceBundle::LOAD_COMMON_RESOURCES);
+      ui::ResourceBundle::GetSharedInstance().AddDataPackFromAsset("assets/resources.pak");
+    }
```
Please add `#else` part of `#if defined(CASTANETS)` here.

venumusham replied:
will add else part, for default case

Sunwoo-Nam commented on chrome/app/chrome_main_delegate.cc:
```diff
@@ -851,11 +855,13 @@ void ChromeMainDelegate::PreSandboxStartup() {
-    ui::ResourceBundle::GetSharedInstance().AddDataPackFromAsset("assets/resources.pak");
+    if (base::Castanets::IsEnabled()) {
+      ui::MaterialDesignController::Initialize();
+      const std::string loaded_locale =
```
loaded_locale is being used out of this scope. (line 911)
It can make build break during Android build.

venumusham replied:
preparing android build, will handle this

Sunwoo-Nam commented on components/discardable_memory/client/client_discardable_shared_memory_manager.cc:
```diff
@@ -361,8 +368,28 @@ ClientDiscardableSharedMemoryManager::AllocateLockedDiscardableSharedMemory(
-  if (!memory->CreateAndMap(size))
-    base::TerminateBecauseOutOfMemory(size);
+  if (base::Castanets::IsEnabled()) {
+    memory = base::MakeUnique<base::DiscardableSharedMemory>();
```
This line is not necessary.

venumusham replied:
IMO, it is needed, AFAIK there was a crash due to this.

Sunwoo-Nam replied:
'memory' has to be initialized two times?

#if defined(CASTANETS)
  auto memory = base::MakeUnique<base::DiscardableSharedMemory>();
  if (base::Castanets::IsEnabled()) {                                                                                                                                                                       
    memory = base::MakeUnique<base::DiscardableSharedMemory>();

venumusham replied:
Yes, memory is defined as auto => to know type of variable we need this definition

Sunwoo-Nam replied:
#if defined(CASTANETS)
  auto memory = **base::MakeUniquebase::DiscardableSharedMemory();** <-- Define & init
  if (base::Castanets::IsEnabled()) {
    memory = **base::MakeUniquebase::DiscardableSharedMemory();**  <--  Is necessary?

venumusham replied:
got your point, multiple initiations are not needed, will handle this, thanks!

Sunwoo-Nam commented on components/viz/client/client_shared_bitmap_manager.cc:
```diff
@@ -186,10 +190,12 @@ void ClientSharedBitmapManager::NotifyRasterizedSharedBitmap(
-  std::vector<uint8_t> pixels_vec(pixels, pixels + memory_size);
-  (*shared_bitmap_allocation_notifier_)
-      ->DidRasterizeSharedBitmap(memory_size, pixels_vec, id);
+  if (base::Castanets::IsEnabled()) {
```
NotifyRasterizedSharedBimap function is called if Castanets is enabled.
I think the runtime flag is not necessary in this function.

venumusham replied:
From cc/resources/video_resource_updater.cc => NotifyRasterizedTile is called even though castanets is disabled, so we need this until Media patches are merged.

Sunwoo-Nam replied:
Okay, I'll make a patch for MM and handle this. :)

venumusham commented:
Review comments are incorporated in new commit, pls review

Sunwoo-Nam commented on mojo/edk/system/node_controller.cc:
```diff
@@ -233,8 +234,9 @@ void NodeController::SendBrokerClientInvitation(
-#if !defined(CASTANETS)
-   DCHECK(!GetConfiguration().is_broker_process);
+#if defined(CASTANETS)
+   if (base::Castanets::IsEnabled())
```
I think this line should be if (!base::Castanets::IsEnabled()).

venumusham replied:
right, will update patchset

venumusham commented:
patchset updated

Sunwoo-Nam reviewed (approved):
LGTM!
